File: spring-cloud-starter/src/main/java/com/jmsoftware/maf/springcloudstarter/quartz/QuartzConfiguration.java
package com.jmsoftware.maf.springcloudstarter.quartz;

import com.jmsoftware.maf.springcloudstarter.property.MafProjectProperties;
import com.jmsoftware.maf.springcloudstarter.quartz.controller.QuartzJobConfigurationController;
import com.jmsoftware.maf.springcloudstarter.quartz.service.QuartzJobConfigurationService;
import com.jmsoftware.maf.springcloudstarter.quartz.service.impl.QuartzJobConfigurationServiceImpl;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import lombok.val;
import org.quartz.Scheduler;
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
import org.springframework.boot.autoconfigure.quartz.QuartzAutoConfiguration;
import org.springframework.boot.autoconfigure.quartz.QuartzProperties;
import org.springframework.boot.autoconfigure.quartz.SchedulerFactoryBeanCustomizer;
import org.springframework.context.annotation.Bean;
import org.springframework.scheduling.quartz.SchedulerFactoryBean;
import org.springframework.transaction.PlatformTransactionManager;

import java.util.Map;
import java.util.Properties;

import static cn.hutool.core.text.CharSequenceUtil.format;
import static org.springframework.scheduling.quartz.SchedulerFactoryBean.PROP_THREAD_COUNT;

/**
 * Description: QuartzConfiguration, change description here.
 *
 * @author Johnny Miller (锺俊), email: johnnysviva@outlook.com, date: 7/13/2021 10:44 PM
 * @see QuartzAutoConfiguration
 **/
@Slf4j
@RequiredArgsConstructor
@ConditionalOnClass({Scheduler.class, SchedulerFactoryBean.class, PlatformTransactionManager.class})
public class QuartzConfiguration {
    private final MafProjectProperties mafProjectProperties;

    @Bean
    public SchedulerFactoryBeanCustomizer schedulerFactoryBeanCustomizer(QuartzProperties quartzProperties) {
        val cpuCoreCount = Runtime.getRuntime().availableProcessors();
        val threadCount = String.valueOf(cpuCoreCount * 2);
        quartzProperties.getProperties().put(PROP_THREAD_COUNT, threadCount);
        return schedulerFactoryBean -> {
            schedulerFactoryBean.setQuartzProperties(this.asProperties(quartzProperties.getProperties()));
            log.warn("Quartz thread pool enhanced by current cpuCoreCount: {}, threadCount: {}", cpuCoreCount,
                     threadCount);
            schedulerFactoryBean.setSchedulerName(
                    format("{}-quartz-scheduler", this.mafProjectProperties.getProjectArtifactId())
            );
            schedulerFactoryBean.setWaitForJobsToCompleteOnShutdown(true);
            schedulerFactoryBean.setBeanName("schedulerFactoryBean");
        };
    }

    private Properties asProperties(Map<String, String> source) {
        val properties = new Properties();
        properties.putAll(source);
        return properties;
    }

    @Bean
    public GreetingBean greetingBean() {
        log.warn("Initial bean: '{}'", GreetingBean.class.getSimpleName());
        return new GreetingBean(this.mafProjectProperties);
    }

    @Bean
    public QuartzJobConfigurationController quartzJobConfigurationController(
            QuartzJobConfigurationService quartzJobConfigurationService
    ) {
        log.warn("Initial bean: '{}'", QuartzJobConfigurationController.class.getSimpleName());
        return new QuartzJobConfigurationController(quartzJobConfigurationService);
    }

    @Bean
    public QuartzJobConfigurationService quartzJobConfigurationService(SchedulerFactoryBean schedulerFactoryBean) {
        log.warn("Initial bean: '{}'", QuartzJobConfigurationServiceImpl.class.getSimpleName());
        return new QuartzJobConfigurationServiceImpl(schedulerFactoryBean, this.mafProjectProperties);
    }
}

File: spring-cloud-starter/src/main/kotlin/com/jmsoftware/maf/springcloudstarter/quartz/QuartzConfiguration.kt
package com.jmsoftware.maf.springcloudstarter.quartz

import com.jmsoftware.maf.common.util.logger
import com.jmsoftware.maf.springcloudstarter.property.MafProjectProperties
import com.jmsoftware.maf.springcloudstarter.quartz.controller.QuartzJobConfigurationController
import com.jmsoftware.maf.springcloudstarter.quartz.service.QuartzJobConfigurationService
import com.jmsoftware.maf.springcloudstarter.quartz.service.impl.QuartzJobConfigurationServiceImpl
import org.quartz.Scheduler
import org.springframework.boot.autoconfigure.condition.ConditionalOnClass
import org.springframework.boot.autoconfigure.quartz.QuartzAutoConfiguration
import org.springframework.boot.autoconfigure.quartz.QuartzProperties
import org.springframework.boot.autoconfigure.quartz.SchedulerFactoryBeanCustomizer
import org.springframework.context.annotation.Bean
import org.springframework.scheduling.quartz.SchedulerFactoryBean
import org.springframework.transaction.PlatformTransactionManager
import java.util.*
import javax.validation.Validator

/**
 * # QuartzConfiguration
 *
 * Description: QuartzConfiguration, change description here.
 *
 * @author Johnny Miller (锺俊), e-mail: johnnysviva@outlook.com, date: 4/15/22 7:19 PM
 * @see QuartzAutoConfiguration
 */
@ConditionalOnClass(Scheduler::class, SchedulerFactoryBean::class, PlatformTransactionManager::class)
class QuartzConfiguration(
    private val mafProjectProperties: MafProjectProperties
) {
    companion object {
        private val log = logger()
    }

    @Bean
    fun schedulerFactoryBeanCustomizer(quartzProperties: QuartzProperties): SchedulerFactoryBeanCustomizer {
        val cpuCoreCount = Runtime.getRuntime().availableProcessors()
        val threadCount = (2 * cpuCoreCount).toString()
        quartzProperties.properties[SchedulerFactoryBean.PROP_THREAD_COUNT] = threadCount
        return SchedulerFactoryBeanCustomizer { schedulerFactoryBean: SchedulerFactoryBean ->
            schedulerFactoryBean.setQuartzProperties(asProperties(quartzProperties.properties))
            log.warn("Quartz thread pool enhanced by current cpuCoreCount: $cpuCoreCount, threadCount: $threadCount")
            schedulerFactoryBean.setSchedulerName("${mafProjectProperties.projectArtifactId}-quartz-scheduler")
            schedulerFactoryBean.setWaitForJobsToCompleteOnShutdown(true)
            schedulerFactoryBean.setBeanName("schedulerFactoryBean")
        }
    }

    private fun asProperties(source: Map<String, String>): Properties {
        val properties = Properties()
        properties.putAll(source)
        return properties
    }

    @Bean
    fun greetingBean(): GreetingBean {
        log.warn("Initial bean: `${GreetingBean::class.java.simpleName}`")
        return GreetingBean(mafProjectProperties)
    }

    @Bean
    fun quartzJobConfigurationController(
        quartzJobConfigurationService: QuartzJobConfigurationService
    ): QuartzJobConfigurationController {
        log.warn("Initial bean: `${QuartzJobConfigurationController::class.java.simpleName}`")
        return QuartzJobConfigurationController(quartzJobConfigurationService)
    }

    @Bean
    fun quartzJobConfigurationService(schedulerFactoryBean: SchedulerFactoryBean, validator: Validator): QuartzJobConfigurationService {
        log.warn("Initial bean: `${QuartzJobConfigurationServiceImpl::class.java.simpleName}`")
        return QuartzJobConfigurationServiceImpl(schedulerFactoryBean, mafProjectProperties, validator)
    }
}
